### PR TITLE
dai: support explicit clock control

### PR DESCRIPTION
**Pull request by brentlu**
Add the SOF_IPC_DAI_CLKCTRL command for host to enable/disable DAI
clocks explicitly.

Signed-off-by: Brent Lu <brent.lu@intel.com>

### CODE REVIEW

File: src/drivers/intel/ssp/ssp.c [hunk 1]
```diff
@@ -870,6 +871,31 @@ static int ssp_trigger(struct dai *dai, int cmd, int direction)
...
 }
 
+static int ssp_set_clkctrl(struct dai *dai, struct sof_ipc_dai_clkctrl *clkctrl)
+{
+	int ret = 0;
+
+	dai_info(dai, "ssp_set_clkctrl(): SSP%d clk_en %d clk_id %d",
+		 dai->index, clkctrl->ssp.clk_en, clkctrl->ssp.clk_id);
+
+	/* enable the entire port instead of checking clk_id */
```
Comment (brentlu) [inline src/drivers/intel/ssp/ssp.c:881]: Yeah, I don't know how to control mclk/bclk separately...

File: src/include/sof/lib/dai.h [hunk 2]
```diff
@@ -333,6 +335,20 @@ static inline int dai_set_config(struct dai *dai,
 	return ret;
 }
 
+/**
+ * \brief Digital Audio interface clock control
+ */
+static inline int dai_set_clkctrl(struct dai *dai,
```
Comment (lgirdwood) [inline src/include/sof/lib/dai.h:341]: this can just be called dai_clkctrl()

File: src/include/sof/lib/dai.h [hunk 3]
```diff
@@ -72,6 +73,7 @@ struct sof_ipc_stream_params;
  */
 struct dai_ops {
 	int (*set_config)(struct dai *dai, struct sof_ipc_dai_config *config);
+	int (*set_clkctrl)(struct dai *dai, struct sof_ipc_dai_clkctrl *clkctrl);
```
Comment (lgirdwood) [inline src/include/sof/lib/dai.h:76]: clkctrl

File: src/include/sof/drivers/ssp.h [hunk 4]
```diff
@@ -230,6 +230,7 @@ struct ssp_pdata {
 	uint32_t sscr1;
 	uint32_t psp;
 	uint32_t state[2];		/* SSP_STATE_ for each direction */
+	uint32_t port_ref;
```
Comment (lgirdwood) [inline src/include/sof/drivers/ssp.h:233]: Can we have a comment what this is used for ?

File: src/include/ipc/dai.h [hunk 5]
```diff
@@ -88,4 +88,16 @@ struct sof_ipc_dai_config {
...
 } __attribute__((packed, aligned(4)));
 
+/* general purpose DAI clock control */
+struct sof_ipc_dai_clkctrl {
+	struct sof_ipc_cmd_hdr hdr;
+	uint32_t type;		/**< DAI type - enum sof_ipc_dai_type */
+	uint32_t dai_index;	/**< index of this type dai */
+
+	/* HW specific data */
+	union {
```
Comment (lgirdwood) [inline src/include/ipc/dai.h:98]: Dont think we need the union here as the interface looks generic.

Comment (lgirdwood) [reply src/include/ipc/dai.h:98]: in fact, I would just put the
```c
	uint32_t clk_en;	/**&lt; Zero to disable the clock */
        uint32_t clk_id;	/**&lt; Clock id to operate */
```
in here directly

File: src/include/ipc/dai-intel.h [hunk 6]
```diff
@@ -205,4 +215,12 @@ struct sof_ipc_dai_dmic_params {
 	struct sof_ipc_dai_dmic_pdm_ctrl pdm[SOF_DAI_INTEL_DMIC_NUM_CTRL];
 } __attribute__((packed, aligned(4)));
 
+/* SSP Clock Control Request - SOF_IPC_DAI_SSP_CLKCTRL */
+struct sof_ipc_dai_ssp_clkctrl {
```
Comment (lgirdwood) [inline src/include/ipc/dai-intel.h:219]: dont think this struct is needed.

File: src/include/ipc/dai-intel.h [hunk 7]
```diff
@@ -60,6 +60,16 @@
 /* DMIC max. four controllers for eight microphone channels */
 #define SOF_DAI_INTEL_DMIC_NUM_CTRL			4
 
+/* SSP clocks to enable/disable
+ *
+ * Macros for clk_id field in sof_ipc_dai_ssp_clkctrl struct.
+ */
+
+/* mclk */
+#define SOF_DAI_INTEL_SSP_CLKCTRL_MCLK			1
```
Comment (lgirdwood) [inline src/include/ipc/dai-intel.h:69]: lets keep it generic and remove "SSP" from these names.

File: src/drivers/intel/ssp/ssp.c [hunk 8]
```diff
@@ -627,11 +627,11 @@ static int ssp_pre_start(struct dai *dai)
 
 	int ret = 0;
 
-	dai_info(dai, "ssp_pre_start()");
+	dai_info(dai, "ssp_port_start(), port_ref = %d", ssp->port_ref);
 
-	/* SSP active means bclk already configured. */
-	if (ssp->state[SOF_IPC_STREAM_PLAYBACK] == COMP_STATE_ACTIVE ||
-	    ssp->state[SOF_IPC_STREAM_CAPTURE] == COMP_STATE_ACTIVE)
+	ssp->port_ref++;
```
Comment (lgirdwood) [inline src/drivers/intel/ssp/ssp.c:632]: Can we not use "state" here, I'm assuming this needs to be in a configured but inactive state ?

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Best to get inputs from @plbossart, @keyonjie and @slawblauciak

**Comment by lgirdwood** [comment]
@bardliao @keyonjie @RanderWang @mengdonglin fyi - please lets all help on the solution.
@brentlu do you have any kernel PR ?

**Comment by brentlu** [comment]
Kernel PR is here:
https://github.com/thesofproject/linux/pull/2932

File: src/drivers/intel/ssp/ssp.c (hunk 1 above)
Comment (bardliao) [reply src/drivers/intel/ssp/ssp.c:881]: I would suggest that we add two helper functions to control mclk and bclk respectively. We count ref count in those helper functions. And we use the helper functions within` ssp_set_clkctrl`, `ssp_port_stop`, and `ssp_port_start`.

**Review by bardliao — COMMENTED** [review]
Thanks @brentlu for starting this. See my comments below

**Comment by brentlu** [comment]
Thanks @bardliao, we still need to figure our how to enable/disable mclk/bclk separately. In this test patch I just enable entire port and check the log to see if the IPC command works.

File: src/include/sof/lib/dai.h (hunk 2 above)
Comment (brentlu) [reply src/include/sof/lib/dai.h]: fixed

File: src/include/sof/lib/dai.h (hunk 3 above)
Comment (brentlu) [reply src/include/sof/lib/dai.h]: fixed

File: src/include/ipc/dai.h (hunk 5 above)
Comment (brentlu) [reply src/include/ipc/dai.h]: fixed

File: src/include/ipc/dai-intel.h (hunk 6 above)
Comment (brentlu) [reply src/include/ipc/dai-intel.h]: fixed

File: src/include/ipc/dai-intel.h (hunk 7 above)
Comment (brentlu) [reply src/include/ipc/dai-intel.h]: fixed

File: src/drivers/intel/ssp/ssp.c (hunk 8 above)
Comment (brentlu) [reply src/drivers/intel/ssp/ssp.c:632]: the state checking is added back.

**Comment by plbossart** [comment]
@brentlu @lgirdwood I don't think it's required to add new IPCs for the current version of SOF, we can get by with what @bardliao is working on. It'd rather add this capability at a later stage when we have fewer in-flight IPC transitions on-going.

**Comment by plbossart** [comment]
closing since we've enabled the early bitclock support.

**Pull request closed without merging**